syntax = "proto3";

/*TaskController*/
service Task {
    /*
    创建一个训练任务
    return {guid}
    */
    rpc createTrain (stream FileBody) returns (Response);
    /*
    获取训练模型
    return {
        bert_config_file
        vocab_file
        tags_file
        bert_model
        lstm_crf_model
    }
    */
    rpc getModel (GetInfo) returns (stream FileBody);
    /*
    创建一个预测任务
    return {guid}
    */
    rpc createPredict (stream FileBody) returns (Response);
}

/*TrainController*/
service Train {
    rpc train (TrainInfo) returns (stream Response);
}

/*PredictController*/
service Predict {
    rpc pred (PredictInfo) returns (Response);
}


message TrainInfo {
    string guid = 1;
    int32 num_epochs = 2;
    repeated int32 num_gpus = 3;
    int32 hidden_dim = 4;
    string word_tag_split = 5;
    string pattern = 6;
    int32 padding_length = 7;
    int32 batch_size = 8;
    int32 eval_batch_size = 9;
}

message PredictInfo {
    string guid = 1;
    bool use_gpu = 2;
    int32 hidden_dim = 3;
}

message GetInfo {
    string guid = 1;
    string model_id = 2;
}

message Response {
    string status = 1;
    string result = 2;
}

message FileBody {
    string name = 1;
    bytes file = 2;
}